feat: Enhances breakpoint editing

blehrer opened the pull request:
The keymapping `<leader>B` (in `kickstart.plugins.debug`) is now configured to guide users through the process of adding a `condition`, `hitCondition`, and `logMessage` to a breakpoint.

This also allows attributes that have already been set to be edited, instead of having to start from scratch each time you want to update one of these fields.

These breakpoints will persist for the duration of the session.

Tested with  v0.11.1

blehrer committed:
feat: Enhances breakpoint editing

The keymapping `<leader>B` is now configured to guide users through the
process of adding a `condition`, `hitCondition`, and `logMessage` to a
breakpoint.

blehrer commented:
small note to the reviewers: `h:dap.toggle_breakpoint()` misspells variable names. Compare [docs][docs] to [src][src].

[docs]: https://github.com/mfussenegger/nvim-dap/blob/8df427aeba0a06c6577dc3ab82de3076964e3b8d/doc/dap.txt#L795-L809
[src]: https://github.com/mfussenegger/nvim-dap/blob/8df427aeba0a06c6577dc3ab82de3076964e3b8d/lua/dap/protocol.lua#L295-L301

blehrer committed:
refactor to use no magic numbers

blehrer committed:
revert testing change

oriori1703 commented on lua/kickstart/plugins/debug.lua:
```diff
@@ -64,9 +64,67 @@ return {
       '<leader>B',
       function()
-        require('dap').set_breakpoint(vim.fn.input 'Breakpoint condition: ')
+        require 'dap.protocol'
```
Why is this needed?

blehrer replied:
It is where `dap.SourceBreakpoint` is declared. This is used to describe the return type of `find_bp()`. It might be more idiomatic to use a `@module` ... but also, it seems like my lsp can find the definition of this type without doing so. I will remove it.

oriori1703 replied:
I noticed the same :)

oriori1703 commented on lua/kickstart/plugins/debug.lua:
```diff
@@ -64,9 +64,67 @@ return {
+        require 'dap.protocol'
+        local dap = require 'dap'
+        -- Search for an existing breakpoint on this line in this buffer
+        ---@return dap.SourceBreakpoint bp that was either found, or an empty placeholder
+        local function find_bp()
+          local buf_bps = require('dap.breakpoints').get(vim.fn.bufnr())[vim.fn.bufnr()]
+          ---@type dap.SourceBreakpoint
+          local bp = { condition = '', logMessage = '', hitCondition = '', line = vim.fn.line '.' }
+          for _, candidate in ipairs(buf_bps) do
+            if candidate.line and candidate.line == vim.fn.line '.' then
+              bp = candidate
+              break
+            end
+          end
+          return bp
+        end
```
This simplifies the code a bit
```suggestion
        local dap = require 'dap'

        -- Search for an existing breakpoint on this line in this buffer
        ---@return dap.SourceBreakpoint bp that was either found, or an empty placeholder
        local function find_bp()
          local buf_bps = require('dap.breakpoints').get(vim.fn.bufnr())[vim.fn.bufnr()]
          ---@type dap.SourceBreakpoint
          for _, candidate in ipairs(buf_bps) do
            if candidate.line and candidate.line == vim.fn.line '.' then
              return candidate
            end
          end

          return { condition = '', logMessage = '', hitCondition = '', line = vim.fn.line '.' }
        end

```

oriori1703 commented on lua/kickstart/plugins/debug.lua:
```diff
@@ -64,9 +64,67 @@ return {
+          end
+          vim.ui.select(menu_options, {
+            prompt = 'Edit Breakpoint',
+          }, function(choice)
```
This should solve an error that would happen if a user cancelled the dialog 

```suggestion
          }, function(choice)
            if choice == nil then
              -- User cancelled the selection
              return
            end
```

oriori1703 commented on lua/kickstart/plugins/debug.lua:
```diff
@@ -64,9 +64,67 @@ return {
+          vim.ui.select(menu_options, {
+            prompt = 'Edit Breakpoint',
+          }, function(choice)
+            local prompt = (tostring(choice)):gsub(':.*', '')
```
for a cleaner implementation, you should use the `format_item` opt.
This way you don't need to remove the formatting from the user's choice.

i.e. something like this:
```suggestion
          for k, _ in pairs(props) do
            table.insert(menu_options, k)
          end
          vim.ui.select(menu_options, {
            prompt = 'Edit Breakpoint',
            format_item = function(item) return ('%s: %s'):format(item, props[item].value) end,
          }, function(choice)
```

oriori1703 reviewed (changes requested):
This looks a like a great addition 😄 

I commented about a few suggestions / fixes

blehrer committed:
lazily init empty breakpoint

Instead of creating an empty object to fill, only create one if no match is found.

Co-authored-by: Ori Perry <48057913+oriori1703@users.noreply.github.com>

blehrer committed:
declare vim.ui.select opt `format_item`

Co-authored-by: Ori Perry <48057913+oriori1703@users.noreply.github.com>

blehrer committed:
handle cancellation

Co-authored-by: Ori Perry <48057913+oriori1703@users.noreply.github.com>

blehrer committed:
Remove unnecessary import

blehrer committed:
fix styleua issue

oriori1703 commented on lua/kickstart/plugins/debug.lua:
```diff
@@ -64,9 +64,72 @@ return {
+            end
+            props[prompt].setter(vim.fn.input {
+              prompt = ('[%s] '):format(prompt),
+              default = props[prompt].value,
```
After the PR changes `prompt` is not defined anymore.
This also happened to me when I adapted your changes to my fork 😆 

```suggestion
              props[choice].setter(vim.fn.input {
              prompt = ('[%s] '):format(choice),
              default = props[choice].value,

```

blehrer replied:
i ran into the same thing.... currently working on a fix...

blehrer replied:
or... i'll just merge yours ;)

blehrer committed:
fix merge issue

Co-authored-by: Ori Perry <48057913+oriori1703@users.noreply.github.com>

blehrer committed:
stylua

oriori1703 reviewed (approved):
LGTM 😄

oriori1703 committed:
Fix stylua formatting
